pinot-common/src/main/java/org/apache/pinot/common/function/scalar/JsonFunctions.java

```diff
@@ -321,17 +326,296 @@ private static void setValuesToMap(String keyColumnName, String valueColumnName,
+        for (int i = 0; i < node.size(); i++) {
+          String path = "$[" + i + "]";
+          if (dotNotation) {
+            keys.add(String.valueOf(i));
+          } else {
+            keys.add(path);
+          }
+        }
+      }
+    } else if (jsonPath.matches("\\$\\.[^.]+\\.\\*")) {
+      // Pattern like $.field.*
+      String fieldPath = jsonPath.substring(2, jsonPath.length() - 2); // Remove $. and .*
+      JsonNode targetNode = node.get(fieldPath);
+      if (targetNode != null) {
+        if (targetNode.isObject()) {
+          targetNode.fieldNames().forEachRemaining(fieldName -> {
+            String path = "$['" + fieldPath + "']['" + fieldName + "']";
+            if (dotNotation) {
+              keys.add(fieldPath + "." + fieldName);
+            } else {
+              keys.add(path);
+            }
+          });
+        } else if (targetNode.isArray()) {
+          for (int i = 0; i < targetNode.size(); i++) {
+            String path = "$['" + fieldPath + "'][" + i + "]";
+            if (dotNotation) {
+              keys.add(fieldPath + "." + i);
+            } else {
+              keys.add(path);
+            }
+          }
+        }
+      }
+    }
+    return keys;
+  }
+
+  /**
+   * Extract all keys recursively from a JSON object up to maxDepth
+   */
+  private static List<String> extractAllKeysRecursively(Object jsonObj, int maxDepth) {
+    return extractAllKeysRecursively(jsonObj, maxDepth, false);
+  }
+
+  /**
+   * Extract all keys recursively from a JSON object up to maxDepth with output format option
+   */
+  private static List<String> extractAllKeysRecursively(Object jsonObj, int maxDepth, boolean dotNotation) {
+    List<String> allKeys = new java.util.ArrayList<>();
+    try {
+      JsonNode node;
+      if (jsonObj instanceof String) {
+        node = JsonUtils.stringToJsonNode((String) jsonObj);
+      } else {
+        node = JsonUtils.stringToJsonNode(JsonUtils.objectToString(jsonObj));
+      }
+
+      extractKeysFromNode(node, "$", allKeys, maxDepth, 1, dotNotation);
+    } catch (Exception e) {
+      // Return empty list on error
+    }
+    return allKeys;
+  }
+
+  /**
+   * Recursively extract keys from a JsonNode
+   */
+  private static void extractKeysFromNode(JsonNode node, String currentPath, List<String> keys,
+      int maxDepth, int currentDepth) {
+    extractKeysFromNode(node, currentPath, keys, maxDepth, currentDepth, false);
+  }
+
+  /**
+   * Recursively extract keys from a JsonNode with output format option
+   */
+  private static void extractKeysFromNode(JsonNode node, String currentPath, List<String> keys,
+      int maxDepth, int currentDepth, boolean dotNotation) {
+    if (currentDepth > maxDepth) {
+      return;
+    }
+
+    if (node.isObject()) {
+      node.fieldNames().forEachRemaining(fieldName -> {
+        String newPath = currentPath + "['" + fieldName + "']";
+        String keyToAdd = dotNotation ? convertToDotNotation(newPath) : newPath;
```

Comment:
When dot notation is enabled, can we also pass the key so that we don't need to do regexp replace? We can directly construct the dot notation key